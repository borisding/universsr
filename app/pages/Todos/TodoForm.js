import React, { useState } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { TextInput } from '@app/common/components';
import './TodoForm.module.scss';

const initialState = {
  isInvalid: false,
  todoInput: ''
};

export default function TodoForm(props) {
  const [isInvalid, setIsInvalid] = useState(initialState.isInvalid);
  const [todoInput, setTodoInput] = useState(initialState.todoInput);

  const styles = classNames('todo-input', { 'error-input': !!isInvalid });
  const placeholder = isInvalid ? 'Todo is required!' : 'Enter new todo...';

  function handleChange(evt) {
    setIsInvalid(initialState.isInvalid);
    setTodoInput(evt.target.value);
  }

  function handleBlur(evt) {
    if (isInvalid) {
      setIsInvalid(!isInvalid);
      setTodoInput(evt.target.value);
    }
  }

  function handleSubmit(evt) {
    evt.preventDefault();

    if (todoInput.trim() === '') {
      setIsInvalid(!initialState.isInvalid);
      setTodoInput(initialState.todoInput);
    } else {
      props.addTodo(todoInput);
      setIsInvalid(initialState.isInvalid);
      setTodoInput(initialState.todoInput);
    }
  }

  return (
    <form onSubmit={handleSubmit}>
      <TextInput
        name="newTodo"
        value={todoInput}
        styleName={styles}
        placeholder={placeholder}
        onChange={handleChange}
        onBlur={handleBlur}
      />
    </form>
  );
}

TodoForm.propTypes = {
  addTodo: PropTypes.func.isRequired
};
