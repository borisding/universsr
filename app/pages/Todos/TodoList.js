import React from 'react';
import styles from './TodoList.module.scss';

export default function TodoList({ todos }) {
  const totalCompleted = todos.filter(todo => !!todo.completed).length;
  return (
    <div className={styles.todosContainer}>
      <h3>{`Todo List (${totalCompleted} / ${todos.length})`}</h3>
      <ul>
        {todos.map(todo => (
          <li key={todo.id}>
            <span
              className={!todo.completed ? styles.pending : styles.completed}
            >
              {todo.title} (ID: {todo.id})
            </span>
          </li>
        ))}
      </ul>
    </div>
  );
}
