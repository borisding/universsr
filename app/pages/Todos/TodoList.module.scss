@import '../../common/styles/variables';

.todosContainer {
  ul {
    list-style: none;
  }

  ul > li {
    padding-bottom: $base-padding / 3;
    padding-top: $base-padding / 3;
  }

  span {
    display: inline-block;
  }

  .pending {
    color: $color-blue;
  }

  .completed {
    color: $color-black;
    text-decoration: line-through;
  }
}
