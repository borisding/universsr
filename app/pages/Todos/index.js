import lazyLoad from '../lazyLoad';

export default lazyLoad(() =>
  import(/* webpackChunkName: 'todos' */ './Todos')
);
