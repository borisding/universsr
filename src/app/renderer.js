import React from 'react';
import DocumentTitle from 'react-document-title';
import serialize from 'serialize-javascript';
import flushChunks from 'webpack-flush-chunks';
import { renderToString } from 'react-dom/server';
import { flushChunkNames } from 'react-universal-component/server';
import { matchRoutes, renderRoutes } from 'react-router-config';
import { StaticRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import storeFactory from './store';
import routes from './routes';

// preload data for matched route
function prefetchBranchData(store, url) {
  try {
    const branch = matchRoutes(routes, url);
    const promises = branch.map(({ route, match }) => {
      const { loadData } = route;
      const { dispatch } = store;

      if (match && match.isExact && loadData) {
        return Array.isArray(loadData)
          ? Promise.all(loadData.map(action => dispatch(action(match))))
          : dispatch(loadData(match));
      }

      return Promise.resolve(null);
    });

    return Promise.all(promises);
  } catch (err) {
    throw err;
  }
}

// export default server renderer and receiving stats
// also, should allow it to be mounted as middleware for production usage
export default function serverRenderer({ clientStats }) {
  return async (req, res, next) => {
    try {
      const context = {};
      const store = storeFactory();

      await prefetchBranchData(store, req.url);

      const appString = renderToString(
        <Provider store={store}>
          <StaticRouter location={req.url} context={context}>
            {renderRoutes(routes)}
          </StaticRouter>
        </Provider>
      );

      const pageTitle = DocumentTitle.rewind();
      const chunksOptions = { chunkNames: flushChunkNames() };
      const { scripts, stylesheets } = flushChunks(clientStats, chunksOptions);

      const preloadedState = serialize(store.getState(), { isJSON: true });
      const { statusCode = 200, redirectUrl } = context;
      const { nonce } = res.locals;

      if ([301, 302].includes(statusCode) && redirectUrl) {
        return res.redirect(statusCode, redirectUrl);
      }

      res.status(statusCode);
      res.render('index', {
        pageTitle,
        appString,
        preloadedState,
        stylesheets,
        scripts,
        nonce
      });
    } catch (err) {
      next(new Error(err));
    }
  };
}
