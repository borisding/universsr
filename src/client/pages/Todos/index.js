import lazy from '@client/pages/lazy';

export default lazy(() => import(/* webpackChunkName: 'todos' */ './Todos'));
