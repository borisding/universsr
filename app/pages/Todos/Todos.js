import React from 'react';
import { useFrontload } from 'react-frontload';
import { Page, Loader } from '../../components';
import TodoList from './TodoList';

export default function Todos() {
  const { data: todos, frontloadMeta } = useFrontload(
    'todos',
    async ({ api }) => {
      const todos = await api.fetchTodos();
      return todos;
    }
  );

  if (frontloadMeta.pending) {
    return <Loader />;
  }

  return (
    <Page title="Todos">
      <TodoList todos={todos} />
    </Page>
  );
}
