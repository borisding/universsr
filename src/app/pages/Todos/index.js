import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { PageTitle, Loader } from '@common/components';
import TodoForm from './TodoForm';
import TodoList from './TodoList';
import * as todoActions from './actions';

class Todos extends Component {
  static propTypes = {
    todos: PropTypes.arrayOf(PropTypes.object).isRequired,
    actions: PropTypes.objectOf(PropTypes.func).isRequired
  };

  componentDidMount() {
    this.props.actions.prefetchTodos();
  }

  renderTodos = () => {
    const { isFetching, actions, todos } = this.props;

    if (isFetching) {
      return <Loader />;
    }

    return (
      <Fragment>
        <TodoForm addTodo={actions.addTodo} />
        <hr />
        <TodoList todos={todos} updateTodo={actions.updateTodo} />
      </Fragment>
    );
  };

  render() {
    return (
      <PageTitle title="Todos">
        <h3>Todos Demo</h3>
        {this.renderTodos()}
      </PageTitle>
    );
  }
}

export default connect(
  state => ({ todos: state.todos.data, isFetching: state.todos.isFetching }),
  dispatch => ({ actions: bindActionCreators(todoActions, dispatch) })
)(Todos);
