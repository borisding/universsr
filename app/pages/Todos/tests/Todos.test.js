import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import Todos from '../Todos';

jest.mock('react-frontload', () => ({
  useFrontload: () => ({ data: [], frontloadMeta: {} })
}));

describe('Todos', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <Todos />
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
