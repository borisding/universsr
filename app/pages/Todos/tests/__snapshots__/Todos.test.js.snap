// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todos renders correctly 1`] = `
<div
  className="todosContainer"
>
  <h3>
    Todo List (0 / 0)
  </h3>
  <ul />
</div>
`;
